Track min and max of important quantities

baperry2 opened the pull request:
I made some modifications so the terminal output and datalog files track extrema of important quantities (e.g., species mass fractions) for debugging purposes and thought it might be generally useful. This tracking occurs at the same intervals as tracking integrated quantities but can be turned off independently. 

Some notes:
- I modified the function `maxDerive`, which in the previous implementation ignored coarse cells covered by fine cells (good), but assumed maxima were positive (bad, e.g. for eint). The new implementation does not ignore covered cells, but this should be okay - average down is always called before the tracking, so covered cells should never have more extreme values than the cells covering them.
- "massfrac" reports extrema across all mass fractions. The individual extrema for the species mass fractions corresponding to `pelec.fuel_name`, `pelec.ftrac_spec_name`, and a new input called `pelec.extrema_spec_name` are also reported if these input parameters are specified.

marchdf commented:
Hey Bruce, this is great! There are a bunch of things the linter picked up if you wouldn't mind fixing those. I left some other minor comments

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+void
+PeleC::monitor_extrema()
+{
+  BL_PROFILE("PeleC::sum_integrated_quantities()");
```
rename the BL_PROFILE stuff

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+  bool local_flag = true;
+
+  int finest_level = parent->finestLevel();
+  amrex::Real time = state[State_Type].curTime();
```
const this

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+
+  int nspec_extrema = 1;
+  if (!fuel_name.empty()) {
+    nspec_extrema += 1;
```
`++` instead of `+=1` (here and below)

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+  auto nextrema = extrema_vars.size();
+  constexpr amrex::Real neg_huge = std::numeric_limits<amrex::Real>::lowest();
+  constexpr amrex::Real huge = std::numeric_limits<amrex::Real>::max();
+  amrex::Vector<amrex::Real> minima, maxima;
```
the next four or five lines can just be put in the construction: `minima(nextrema, huge)`, basically the same thing for `maxima` too.

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+
+  for (int lev = 0; lev <= finest_level; lev++) {
+    PeleC& pc_lev = getLevel(lev);
+    for (int ii = 0; ii < nextrema - nspec_extrema; ++ii) {
```
why `ii` and not just `i`?

baperry2 replied:
I dislike single character variable names and try to avoid them to make search, find+replace, etc, easier to use.

marchdf replied:
lol

jrood-nrel replied:
Is your editor the stickies app?

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+    return;
+  }
+
+  bool local_flag = true;
```
const this?

marchdf commented on Source/SumIQ.cpp:
```diff
@@ -138,3 +138,137 @@ PeleC::sum_integrated_quantities()
+    auto mf = derive("massfrac", time, 0);
+    BL_ASSERT(!(mf == nullptr));
+    int idx_massfrac = nextrema - nspec_extrema;
+    bool local = true;
```
const this? Or take the same as above `local_flag`?

marchdf commented:
Is there a reason not to just output the extrema of all the mass fractions? I guess it can get verbose...

marchdf commented:
And can you post an example of the log file with these modifications? Curious to see what it looks like.

baperry2 commented:
Here is what the output looks like:
![Screen Shot 2021-11-17 at 9 34 23 AM](https://user-images.githubusercontent.com/53018946/142242020-9e6d6d05-2722-410e-bf26-c861b1ef8fda.png)

And here is the log file:
![Screen Shot 2021-11-17 at 9 34 03 AM](https://user-images.githubusercontent.com/53018946/142242047-5503bf8c-3e81-40f1-b252-47bbcb7ac799.png)

jrood-nrel commented:
I think I would be fine replacing the integrated quantities with this.

shashankNREL commented:
@baperry2 : A small suggestion - can you either print on the screen as exponentials with only a few digits or print just the one decimal point. We perhaps don't need that much resolution in min and max. You can retain the full in the log file.

shashankNREL commented:
> @baperry2 : A small suggestion - can you either print on the screen as exponentials with only a few digits or print just the one decimal point. We perhaps don't need that much resolution in min and max. You can retain the full in the log file.

marchdf commented:
> I think I would be fine replacing the integrated quantities with this.

I want to keep the integrated quantities. I use them for things like TG and quick plotting of integral quantities.

marchdf commented:
> @baperry2 : A small suggestion - can you either print on the screen as exponentials with only a few digits or print just the one decimal point. We perhaps don't need that much resolution in min and max. You can retain the full in the log file.

I second this. Can we also make it look more like the other ones instead of whitespace before the name, make the whitespace between the name and the equal? Also, a space between `MIN/MAX` and `=` sign?

baperry2 commented:
Okay, I think I addressed all of the above comments. I also added tracking for the sum of mass fractions (actually sumY - 1 to make any discrepancy easier to track without requiring crazy precision). And I added an option to print extrema for all mass fractions by setting `pelec.extrema_spec_name = ALL`.

marchdf commented:
Excellent idea. Can you repost the new output here?

baperry2 commented:
Log file:
![Screen Shot 2021-11-17 at 1 26 58 PM](https://user-images.githubusercontent.com/53018946/142277091-e30babc0-4840-430c-9aad-f9c7cc0517b9.png)

Output:
![Screen Shot 2021-11-17 at 1 26 18 PM](https://user-images.githubusercontent.com/53018946/142277119-9efa45bf-0fbb-482c-b9a7-f4dfb2bf0c78.png)

(without the option to print all species to keep these reasonably sized)

marchdf commented:
Sweet. One last thing. Can you remove `TEMP` from the Integrated quantities ?

marchdf commented on Source/Params/_cpp_parameters:
```diff
@@ -396,10 +396,16 @@ track_grid_losses            int            0
 sum_per                      Real          -1.0e0
 
-# abort if we exceed CFL = 1 over the cource of a timestep
+# abort if we exceed CFL = 1 over the course of a timestep
```
coarse not course

jrood-nrel commented:
Do we need to list `TIME` before everything? Looks really redundant now.

drummerdoc commented:
Not required, but makes it easy to make plots using grep in the log

jrood-nrel commented:
Maybe we could just use `QUANTITY` or something.

baperry2 commented:
@drummerdoc I've found having TIME in the output handy for the same reason. But the datalog file is there for plotting purposes (except on runs when you didn't set it up, which you sometimes still want to plot).

My preference would be to record datalog files by default.

I need to reshuffle the log file in some way because I noticed it conflicts with the prob log file for some of the test cases.

baperry2 commented:
Ok, I removed `TEMP` but left `TIME` there. Also modified things so datalogs are specified by name instead of index to avoid potential overlap. I think this is ready to merge.

marchdf commented:
there's even documentation!!
